plugins/hls-gadt-plugin/src/Ide/Plugin/GADT.hs

```diff
@@ -0,0 +1,135 @@
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { Ide.Types.pluginHandlers =
+        mkPluginHandler STextDocumentCodeAction codeActionHandler
+    , pluginCommands =
+        [PluginCommand toGADTSyntaxCommandId "convert data decl to GADT syntax" toGADTCommand]
+    }
+
+-- | Parameter used in the command
+data ToGADTParams = ToGADTParams
+    { uri          :: Uri
+    , range        :: Range
+    -- PluginId doesn't have instance of ToJSON and FromJSON
+    , pluginIdText :: T.Text
+    } deriving (Generic, ToJSON, FromJSON)
+
+toGADTSyntaxCommandId :: CommandId
+toGADTSyntaxCommandId = "GADT.toGADT"
+
+-- | A command replaces H98 data decl with GADT decl in place
+toGADTCommand :: CommandFunction IdeState ToGADTParams
+toGADTCommand state ToGADTParams{..} = response $ do
+    let plId = PluginId pluginIdText
+    nfp <- getNormalizedFilePath plId uri
+    (decls, exts) <- getInRangeH98DeclsAndExts state range nfp
+    (L ann decl) <- handleMaybe
+        ("Expect 1 decl, but got " <> show (Prelude.length decls))
+        (if Prelude.length decls == 1 then Just $ head decls else Nothing)
+    deps <- liftIO $ runAction "GADT.GhcSessionDeps" state $ use GhcSessionDeps nfp
+    (hsc_dflags . hscEnv -> df) <- ExceptT
+        $ pure
+        $ maybeToEither "Get GhcSessionDeps failed" deps
+    txt <- ExceptT $ pure $ T.pack <$> (prettyGADTDecl df . h98ToGADTDecl) decl
+    range <- ExceptT
+        $ pure
+        $ maybeToEither "Unable to get data decl range"
+        $ srcSpanToRange $ locA ann
+    pragma <- getNextPragma state nfp
+    let insertEdit = [insertNewPragma pragma GADTs | all (`notElem` exts) [GADTSyntax, GADTs]]
+
+    _ <- lift $ sendRequest
+            SWorkspaceApplyEdit
+            (ApplyWorkspaceEditParams Nothing (workSpaceEdit nfp (TextEdit range txt : insertEdit)))
+            (\_ -> pure ())
+
+    pure Null
+    where
+        workSpaceEdit nfp edits = WorkspaceEdit
+            (pure $ HashMap.fromList
+                [(filePathToUri $ fromNormalizedFilePath nfp,
+                 List edits)])
+                 Nothing Nothing
+
+codeActionHandler :: PluginMethodHandler IdeState TextDocumentCodeAction
+codeActionHandler state plId@(PluginId txt) (CodeActionParams _ _ doc range _) = response $ do
+    nfp <- getNormalizedFilePath plId doc
+    (inRangeH98Decls, _) <- getInRangeH98DeclsAndExts state range nfp
+    let actions = map (mkAction . printOutputable . tcdLName . unLoc) inRangeH98Decls
+    pure $ List actions
+    where
+        mkAction :: T.Text -> Command |? CodeAction
+        mkAction name = InR CodeAction{..}
+            where
+                _title = "Convert \"" <> name <> "\" to GADT syntax"
+                _kind = Just CodeActionRefactorRewrite
+                _diagnostics = Nothing
+                _isPreferred = Nothing
+                _disabled = Nothing
+                _edit = Nothing
+                _command = Just
```

Comment:
Why are you doing the indirection via a command? You could just compute the edit in the code action handler and return it here!

Reply:
I don't think your command handler is significantly more expensive - I think it also mostly relies on the parsed module, which you also need here.

Reply:
I think users won't use this plugin in most cases, it's unfair to give GADT results every time for these cases.